### ISSUE TITLE
Transactions Failing in Lambda

### ISSUE THREAD

**Issue by zeldar** [dynamoosejs/dynamoose#695]
### Summary:
I'm using Dynamoose in Node lambdas with version 10.x.  I get this error when trying to use transactions. 

It works on my local machine with serverless offline but not in a lambda.

TypeError: Cannot read property 'transactWriteItems' of undefined
    at transact (/var/task/node_modules/dynamoose/dist/Dynamoose.js:226:26)
    at Dynamoose.transaction (/var/task/node_modules/dynamoose/dist/Dynamoose.js:262:13)
    at process._tickCallback (internal/process/next_tick.js:68:7)

#### Schema
```
// Code here
```

### Current output and behavior:

TypeError: Cannot read property 'transactWriteItems' of undefined
    at transact (/var/task/node_modules/dynamoose/dist/Dynamoose.js:226:26)
    at Dynamoose.transaction (/var/task/node_modules/dynamoose/dist/Dynamoose.js:262:13)
    at process._tickCallback (internal/process/next_tick.js:68:7)

### Expected output and behavior:

For the transaction to be carried out.

### Environment:
Lambda Nodejs10.x

Operating System:
Operating System Version:
Node.js version (`node -v`):
NPM version: (`npm -v`):
Dynamoose version:
Dynamoose Plugins:

-

### Dynamoose Plugins:

- [ ] Yes, I believe that one or multiple 3rd party Dynamoose plugins are effecting this issue
- [ X] No, I believe this issue is independent of any 3rd party Dynamoose plugins I'm using
- [ ] Unknown, I'm unsure if Dynamoose plugins are effecting this issue
- [X ] I am not using any Dynamoose plugins

### Type (select 1):
- [X ] Bug report
- [ ] Feature suggestion
- [X ] Question
- [ ] Other suggestion
- [ ] Something not listed here

### Other:
- [X ] I have read through the Dynamoose documentation before posting this issue
- [X ] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [X ] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [X ] I have tested the code provided and am confident it doesn't work as intended
- [X ] I have ensured that all of my plugins that I'm using with Dynamoose are listed above
- [X ] I have filled out all fields above
- [X ] I am running the latest version of Dynamoose

**Comment by fishcharlie** [comment]
Closing this issue for not providing enough detail and not following the issue template. In order to debug this and reproduce it we need a minimal example of this to be able to test it.

**Comment by zeldar** [comment]
@fishcharlie You have never encountered this bug?  I was hoping you may have tried Dynamoose on lambda and seen it happen before.  What information do you need?  I'll update the issue the best I can.  I have SO much information, I'm just not sure what to provide you to be useful.

**Comment by zeldar** [comment]
@fishcharlie The fact that this works locally, but not in an actual lambda leads me to believe there is some issue with the lambda environment.  I think if you have a basic lambda in AWS using transactions, you could reproduce this.

**Comment by fishcharlie** [comment]
@zeldar I have not run into this before. You said it works locally but not in AWS Lambda? My biggest question is what your code looks like at this point. I’m assuming it’s failing at a transaction? What does that code look like that you wrote?

Even more helpful would be to narrow down what is failing and provide just that code so I can reproduce it on my own and debug it.

**Comment by zeldar** [comment]
@fishcharlie I updated the original comment with details.  I would have to give you an enormous amount of code to reproduce this exactly, but I have a feeling a basic test case will fail.  

I have seen that transactWriteItems should actually be transactWrite depending on if you are using the DocumentClient or not.

I just want to verify that if you make 2 schemas, and use Model.transaction.create() and then dynamoose.transaction(), that it doesn't fail on lambda for you.

**Comment by zeldar** [comment]
@fishcharlie  Please see this Stack Overflow of a similar issue. I'm wondering if Dynamoose has this issue under the hood.

**Comment by fishcharlie** [comment]
@zeldar Are you using document client tho?

**Comment by zeldar** [comment]
@fishcharlie  here is my set up:

```JavaScript
import dynamoose from 'dynamoose';

dynamoose.AWS.config.update({
  accessKeyId: process.env.ACCESS_KEY_ID,
  secretAccessKey: process.env.SECRET_ACCESS_KEY,
  region: process.env.DYNAMO_REGION,
});

// Defaults to port 8000
if (process.env.STAGE !== 'prod' && process.env.STAGE !== 'dev') {
  dynamoose.local(process.env.DYNAMO_ENDPOINT);
}
const TenantSchema = new dynamoose.Schema(
  {
     ...scheme here
  },
  {
    useNativeBooleans: true,
    useDocumentTypes: true,
    timestamps: true,
  }
);

const Tenant = dynamoose.model(
  `${process.env.STAGE}-${process.env.DATABASE_VERSION}-Tenant`,
  TenantSchema,
  {
    create: false, // Create table in DB, if it does not exist,
    update: false, // Update remote indexes if they do not match local index structure
    waitForActive: false, // Wait for table to be created before trying to use it
    waitForActiveTimeout: 180000, // wait 3 minutes for table to activate
  }
);

export default Tenant;
```

**Comment by fishcharlie** [comment]
@zeldar Can you try using Dynamoose version 1.6.5 and let me know if that fixes your issue?

**Comment by fishcharlie** [comment]
Just for my reference, I think it's failing on the following line.

https://github.com/dynamoosejs/dynamoose/blob/2206935b91f959ed04a6a194ab89c6053fc0f7e1/lib/Dynamoose.ts#L281

I think a possible solution might be to change it to:

```js
this.ddb()[transactionMethodName](transactionReq, async (err, data) => {
```

**Comment by zeldar** [comment]
@fishcharlie Yeah I can try that shortly and then I'll report back and let you know.  Thanks for troubleshooting.

**Comment by zeldar** [comment]
@fishcharlie  That version did not seem to make any difference, but I tried a few things and I think I have narrowed this down to something webpack is doing.  When I run the code locally and check dynamoose.models, all of my models appear as I would expect.  When I do the same thing in my lambda, I get an empty object. 

The new error I get is:  "TransactionError: dev-v2-Tenant is not a registered model. You can only use registered Dynamoose models when using a RAW transaction"

All of my models are set up the same way as what I posted above.  It seems like they register on my local machine when I am running the whole app, but not when they are bundled as individual services.  

Here is my updated webpack config that brought me to this new error:

```JavaScript
const slsw = require('serverless-webpack');
const nodeExternals = require('webpack-node-externals');

module.exports = {
  entry: slsw.lib.entries,
  target: 'node',
  node: {
    __dirname: false,
  },
  // Generate sourcemaps for proper error messages
  devtool: 'source-map',
  // Since 'aws-sdk' is not compatible with webpack,
  // we exclude all node dependencies
  externals: [
    nodeExternals({
      // this WILL include 'aws-sdk'
      whitelist: ['aws-sdk', 'dynamoose'],
    }),
  ],
  mode: slsw.lib.webpack.isLocal ? 'development' : 'production',
  optimization: {
    // We do not want to minimize our code.
    minimize: false,
  },
  performance: {
    // Turn off size warnings for entry points
    hints: false,
  },
  // Run babel on all .js files and skip those in node_modules
  module: {
    rules: [
      {
        test: /\.js$/,
        loader: 'babel-loader',
        include: __dirname,
        exclude: /node_modules/,
      },
    ],
  },
};
```

**Comment by fishcharlie** [comment]
@zeldar Yeah, sorry but I'm not able to run through and help debug that. I'm personally not a fan of build systems like that. They add a lot of complexity and cause the exact problems you are running into now.

At this point tho it doesn't look like this is a Dynamoose issue. If you want you can try my idea of changing that line to use `this.ddb()` instead of `this.dynamoDB`. If you do debug further and determine this is indeed a Dynamoose issue please let me know.

**Comment by zeldar** [comment]
@fishcharlie Just to circle back for closure in case anyone else runs into this issue when creating micro services with webpack.  

The problem was my models lived outside of the directory where my service lived.  Importing dynamoose in two places was causing webpack to have two instances in the app.  One instance that registered the models, and another to call the transaction.  

I fixed this by creating a dbHelper lib that just imports dynamoose, does my env var set up and local checks, and then simply exports the dynamoose instance.  Then, everywhere I imported dynamoose, i imported the dbHelper instead and that solved the issue.

Hopefully this helps someone else.
